test/jdk/com/sun/jndi/dns/ConfigTests/TimeoutWithEmptyDatagrams.java

```diff
@@ -0,0 +1,196 @@
+ * @bug 8339538
+ * @summary Tests that DnsClient correctly calculates left timeout in
+ *          presence of empty datagram packets.
+ * @library ../lib /test/lib
+ * @modules java.base/sun.security.util
+ * @run main/othervm TimeoutWithEmptyDatagrams
+ */
+
+public class TimeoutWithEmptyDatagrams extends DNSTestBase {
+    // initial timeout = 1/4 sec
+    private static final int TIMEOUT = 250;
+    // try 5 times per server
+    private static final int RETRIES = 5;
+    // DnsClient retries again with increased timeout if left
+    // timeout is less than this value, and max retry attempts
+    // is not reached
+    private static final int DNS_CLIENT_MIN_TIMEOUT = 0;
+
+    public TimeoutWithEmptyDatagrams() {
+        setLocalServer(false);
+    }
+
+    public static void main(String[] args) throws Exception {
+        new TimeoutWithEmptyDatagrams().run(args);
+    }
+
+    /*
+     * Tests that we can set the initial UDP timeout interval and the
+     * number of retries.
+     */
+    @Override
+    public void runTest() throws Exception {
+        // Create a DatagramSocket and bind it to the loopback address to simulate
+        // UDP DNS server that doesn't respond
+        try (DatagramSocket ds = new DatagramSocket(new InetSocketAddress(
+                InetAddress.getLoopbackAddress(), 0))) {
+            CountDownLatch gotClientAddress = new CountDownLatch(1);
+            AtomicReference<SocketAddress> clientAddress = new AtomicReference<>();
+            AtomicBoolean stopTestThreads = new AtomicBoolean();
+
+            String allQuietUrl = URIBuilder.newBuilder()
+                    .scheme("dns")
+                    .loopback()
+                    .port(ds.getLocalPort())
+                    .build()
+                    .toString();
+
+            // Run a virtual thread that receives client request packets and extracts
+            // sender address from them.
+            Thread receiverThread = Thread.ofVirtual().start(() -> {
+                while (!stopTestThreads.get()) {
+                    try {
+                        DatagramPacket packet = new DatagramPacket(new byte[1024], 1024);
+                        ds.receive(packet);
+                        System.err.println("Got packet from " + packet.getSocketAddress());
+                        boolean hasClientAddress = clientAddress.get() != null;
+                        clientAddress.set(packet.getSocketAddress());
+                        if (!hasClientAddress) {
+                            gotClientAddress.countDown();
+                        }
+                    } catch (IOException e) {
+                        if (!stopTestThreads.get()) {
+                            throw new RuntimeException(e);
+                        } else {
+                            return;
+                        }
+                    }
+                }
+            });
+
+            // Run a virtual thread that will send an empty packets via server socket
```

Comment:
similarly here the lambda functionality encapsulated in a class abstraction EmptyDatagramSender implements Runnable
Thread.startVirtualThread(new EmptyDatagramSender(. . .))

I make this and the above comment as an observation